### PR TITLE
Create TeleporterMessenger ABI Go binding

### PR DESCRIPTION
**Pull request by gwen917**
## Why this should be merged
- Generate `TeleporterMessenger` ABI Go bindings.
  - ./scripts/abi_go_bindings.sh
  - The script build the `contracts` to generate ABI files, and then generate TeleporterMessenger ABI Go bindings using `abigen`.
- The `abi_go_bindings_checker` github action checks if the generated Go bindings are from the latest contracts

## How this was tested
- CI

### CODE REVIEW

**Comment by geoff-vball** [comment]
Can we make abi_go_bindings.sh generic, so it takes in a file with a list of `.abi.json` files, (and maybe also the desired package name), and then iterates and produces the `.go` abi files for all of them?

**Comment by gwen917** [comment]
> Can we make abi_go_bindings.sh generic, so it takes in a file with a list of `.abi.json` files, (and maybe also the desired package name), and then iterates and produces the `.go` abi files for all of them?

I want to build the contract first, then use the generated files in the `out` dir. In this way, if the contract is changed, we just need one step to generate the `go` files instead of build and then run the script. What if we give a list of contract name, then generate the `go` files of these contracts?

**Comment by geoff-vball** [comment]
Yeah, that'll work!

File: .github/workflows/abi_go_bindings_checker.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  abi_binding:
+    name: abi_binding
+    runs-on: ubuntu-20.04
+
+    steps:
+      - name: Checkout go-ethereum repository
+        uses: actions/checkout@v4
+        with:
+          repository: ethereum/go-ethereum
+          ref: v1.13.1
```
Comment (cam-schultz) [inline .github/workflows/abi_go_bindings_checker.yml:24]: Do we need to checkout Geth? I thought we were using `subnet-evm` `abigen`

File: .github/workflows/check_clean_branch.sh [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+# Exits if any uncommitted changes are found.
+
+set -o errexit
```
Comment (minghinmatthewlam) [inline .github/workflows/check_clean_branch.sh:4]: can you add comments to this file explaining the steps

File: .github/workflows/abi_go_bindings_checker.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
+name: ABI Go Bindings Checker
+
+on:
+  pull_request:
+    branches:
+      - "*"
```
[the remark this answers is not in the document]
Comment (gwen917) [reply .github/workflows/abi_go_bindings_checker.yml:6]: based on Cam's [comments](https://github.com/ava-labs/teleporter/pull/41#discussion_r1344101103), we may not want to run this on `main` in automated fashion

File: .github/workflows/check_clean_branch.sh (hunk 2 above)
Comment (gwen917) [reply .github/workflows/check_clean_branch.sh:4]: added comments.

File: scripts/abi_go_bindings.sh [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+source ./scripts/utils.sh
+
+setARCH
+
+TELEPORTER_PATH=$(
+  cd "$(dirname "${BASH_SOURCE[0]}")"
+  cd .. && pwd
+)
+
+DEFAULT_CONTRACT_LIST="TeleporterMessenger"
```
[the remark this answers is not in the document]
Comment (gwen917) [reply scripts/abi_go_bindings.sh]: added the cross-chain apps. Could you check if I missed anything?

File: .github/workflows/abi_go_bindings_checker.yml (hunk 1 above)
Comment (gwen917) [reply .github/workflows/abi_go_bindings_checker.yml]: yes, removed

File: .github/workflows/check_clean_branch.sh [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+set -o nounset
+set -o pipefail
+
+git update-index --really-refresh >> /dev/null
```
[the remark this answers is not in the document]
Comment (gwen917) [reply .github/workflows/check_clean_branch.sh]: added comments

[the remark this answers is not in the document]
Comment (gwen917) [reply .github/workflows/check_clean_branch.sh]: fixed

File: abis/TeleporterMessenger/TeleporterMessenger.abi.json
[the remark this answers is not in the document]
Comment (gwen917) [reply abis/TeleporterMessenger/TeleporterMessenger.abi.json:1]: removed

File: scripts/abi_go_bindings.sh [hunk 6]
```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env bash
+# Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+# See the file LICENSE for licensing terms.
+
+source ./scripts/utils.sh
+
+setARCH
+
```
[the remark this answers is not in the document]
Comment (gwen917) [reply scripts/abi_go_bindings.sh:10]: added `set -e` so the script will exit at first error.

File: scripts/abi_go_bindings.sh [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    abi_file=$TELEPORTER_PATH/contracts/out/$contract_name.sol/$contract_name.abi.json
+    if ! [ -f $abi_file ]; then
+        echo "Error: Contract $contract_name abi file not found"
+        exit 1
+    fi
+
+    echo "Generating Go bindings for $contract_name..."
+    mkdir -p $TELEPORTER_PATH/abis/$contract_name
+    $GOPATH/bin/abigen --abi $abi_file \
+                       --pkg $(convertToSnakeCase $contract_name) \
```
Comment (geoff-vball) [inline scripts/abi_go_bindings.sh:72]: The naming convention for golang packages is all lowercase with no underscores.

File: .github/workflows/abi_go_bindings_checker.yml [hunk 8]
```diff
@@ -0,0 +1,48 @@
+name: ABI Go Bindings Checker
+
+on:
+  push:
+    branches:
+      - main
```
[the remark this answers is not in the document]
Comment (geoff-vball) [reply .github/workflows/abi_go_bindings_checker.yml]: My understanding is that every workflow should be run in it's own environment. I don't think this test is really only important in PRs though, so I'm fine with this either way.

**Review by geoff-vball — COMMENTED** [review]
Just one small comment here. Otherwise everything looks great.

File: scripts/abi_go_bindings.sh (hunk 7 above)
Comment (gwen917) [reply scripts/abi_go_bindings.sh]: fixed the package name

**Review by geoff-vball — APPROVED** [review]
LGTM!

File: scripts/abi_go_bindings.sh [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+source ./scripts/utils.sh
+
+setARCH
+
+TELEPORTER_PATH=$(
+  cd "$(dirname "${BASH_SOURCE[0]}")"
+  cd .. && pwd
+)
+
+DEFAULT_CONTRACT_LIST="TeleporterMessenger ERC20Bridge ExampleCrossChainMessenger BlockHashPublisher BlockHashReceiver"
```
Comment (cam-schultz) [inline scripts/abi_go_bindings.sh:16]: For completeness, let's also include `BridgeToken`

**Review by cam-schultz — COMMENTED** [review]
Let's include bindings for `BridgeToken.sol`. Otherwise LGTM

File: scripts/abi_go_bindings.sh (hunk 9 above)
Comment (gwen917) [reply scripts/abi_go_bindings.sh]: added `BridgeToken`

**Review by cam-schultz — APPROVED** [review]
LGTM!

**Pull request merged**